src/com/google/javascript/jscomp/CommandLineRunner.java

```diff
@@ -1320,30 +1286,53 @@ private static ConformanceConfig loadConformanceConfig(String configFile)
     Preconditions.checkNotNull(input);
 
     ZipInputStream zip = new ZipInputStream(input);
+    final String envPrefix = options.getEnvironment().toString().toLowerCase()
+        + "/";
+
     Map<String, SourceFile> externsMap = new HashMap<>();
     for (ZipEntry entry = null; (entry = zip.getNextEntry()) != null; ) {
-      BufferedInputStream entryStream = new BufferedInputStream(
-          ByteStreams.limit(zip, entry.getSize()));
-      externsMap.put(entry.getName(),
-          SourceFile.fromInputStream(
-              // Give the files an odd prefix, so that they do not conflict
-              // with the user's files.
-              "externs.zip//" + entry.getName(),
-              entryStream,
-              UTF_8));
+      // Only load externs in the root folder or in a subfolder matching
+      // the specified environment
+      if (entry.getName().indexOf("/") < 0 ||
+          (entry.getName().indexOf(envPrefix) == 0 &&
+              entry.getName().length() > envPrefix.length())) {
+        BufferedInputStream entryStream = new BufferedInputStream(
+            ByteStreams.limit(zip, entry.getSize()));
+        externsMap.put(entry.getName(),
+            SourceFile.fromInputStream(
+                // Give the files an odd prefix, so that they do not conflict
+                // with the user's files.
+                "externs.zip//" + entry.getName(),
+                entryStream,
+                UTF_8));
+      }
     }
 
-    Preconditions.checkState(
-        externsMap.keySet().equals(new HashSet<>(DEFAULT_EXTERNS_NAMES)),
-        "Externs zip must match our hard-coded list of externs.");
-
-    // Order matters, so the resources must be added to the result list
-    // in the expected order.
     List<SourceFile> externs = new ArrayList<>();
-    for (String key : DEFAULT_EXTERNS_NAMES) {
-      externs.add(externsMap.get(key));
+
+    for (String key : BUILTIN_LANG_EXTERNS) {
+      Preconditions.checkState(
+          externsMap.containsKey(key),
+          "Externs zip must contain " + key + ".");
+      externs.add(externsMap.remove(key));
+    }
+
+    // Order matters, so extern resources which have dependencies must be added
+    // to the result list in the expected order.
+    for (String key : BUILTIN_EXTERN_DEP_ORDER) {
+      if (key.indexOf(envPrefix) < 0) {
```

Comment:
The idea was that all extern environments would be added to the same list, but we would only load the ones from the specific folder. Example:

```JavaScript
private static final List<String> BUILTIN_EXTERN_DEP_ORDER = ImmutableList.of(
   //-- Legacy externs --
   "legacy/intl.js",
   "legacy/w3c_event.js",
   "legacy/w3c_event3.js",

   //-- NodeJS exters --
   "nodejs/stream.js",
   "nodejs/path.js"
);
```